agent/xds/clusters.go

```diff
@@ -202,27 +202,38 @@ func (s *Server) clustersFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsho
 
 func (s *Server) clustersFromSnapshotIngressGateway(cfgSnap *proxycfg.ConfigSnapshot) ([]proto.Message, error) {
 	var clusters []proto.Message
-	for _, u := range cfgSnap.IngressGateway.Upstreams {
-		id := u.Identifier()
-		chain, ok := cfgSnap.IngressGateway.DiscoveryChain[id]
-		if !ok {
-			// this should not happen
-			return nil, fmt.Errorf("no discovery chain for upstream %q", id)
-		}
+	createdClusters := make(map[string]bool)
+	for _, upstreams := range cfgSnap.IngressGateway.Upstreams {
+		for _, u := range upstreams {
+			id := u.Identifier()
+
+			// If we've already created a cluster for this upstream, skip it. Multiple listeners may
+			// reference the same upstream, so we don't need to create duplicate clusters in that case.
+			if createdClusters[id] {
+				continue
+			}
 
-		chainEndpoints, ok := cfgSnap.IngressGateway.WatchedUpstreamEndpoints[id]
-		if !ok {
-			// this should not happen
-			return nil, fmt.Errorf("no endpoint map for upstream %q", id)
-		}
+			chain, ok := cfgSnap.IngressGateway.DiscoveryChain[id]
+			if !ok {
+				// this should not happen
+				return nil, fmt.Errorf("no discovery chain for upstream %q", id)
+			}
 
-		upstreamClusters, err := s.makeUpstreamClustersForDiscoveryChain(u, chain, chainEndpoints, cfgSnap)
-		if err != nil {
-			return nil, err
-		}
+			chainEndpoints, ok := cfgSnap.IngressGateway.WatchedUpstreamEndpoints[id]
+			if !ok {
+				// this should not happen
+				return nil, fmt.Errorf("no endpoint map for upstream %q", id)
+			}
+
+			upstreamClusters, err := s.makeUpstreamClustersForDiscoveryChain(u, chain, chainEndpoints, cfgSnap)
+			if err != nil {
+				return nil, err
+			}
 
-		for _, c := range upstreamClusters {
-			clusters = append(clusters, c)
+			for _, c := range upstreamClusters {
+				clusters = append(clusters, c)
+			}
```

Comment:
Not your code, but this could be done with a single append, which I believe will reduce the number of allocations if the size is large:

```go
clusters = append(clusters, upstreamClusters...)
```

Reply:
Went to try this out and ended up running into problems because `clusters` is a `[]proto.Message` and `upstreamClusters` is a `[]*envoy.Cluster`, even though `*envoy.Cluster` implements `proto.Message`. 🤔

Will come back to try and understand this one and see if there is a good way of making the stars align.

Error message:
```
cannot use upstreamClusters (variable of type []*v2.Cluster) as []proto.Message value in argument to append
```